### PR TITLE
dnsdist: update to 1.4.0 w/ new dependencies

### PR DESCRIPTION
**Pull request by James-TR**
Maintainer: me
Compile tested: Ubuntu - OpenWRT buildroot SNAPSHOT
Run tested: Linksys WRT1900ACS armv7 - Snapshot. dnsdist runs correctly, including features that depend on the additional packages included below

Description:
Updates DNSDIST to 1.4.0 which introduced dependencies on two new packages:

libwslay:

  Wslay is a WebSocket library written in C. It implements the protocol version 13 described
   in RFC 6455. This library offers 2 levels of API: event-based API and frame-based low-level
   API. For event-based API, it is suitable for non-blocking reactor pattern style. You can set
   callbacks in various events. For frame-based API, you can send WebSocket frame directly.
   Wslay only supports data transfer part of WebSocket protocol and does not perform opening
   handshake in HTTP.

h2o:

  H2O is a new generation HTTP server that provides quicker response to users with less CPU
   utilization when compared to older generation of web servers. Designed from ground-up, the
   server takes full advantage of HTTP/2 features including prioritized content serving and
   server push, promising outstanding experience to the visitors of your web site.

Signed-off-by: James Taylor <james@jtaylor.id.au>

### CODE REVIEW

**Comment by BKPepe** [comment]
I suggest splitting your single commit into multiple ones. It's preferred to have one single commit per one new package.

File: libs/h2o/Makefile [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+define Package/libh2o
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=H2O Library compiled with libuv
+  URL:=https://h2o.examp1e.net/
+  DEPENDS:=+libuv +libwslay +libopenssl +zlib +libh2o-evloop +libyaml
+endef
+
+define Build/InstallDev
```
Comment (neheb) [inline libs/h2o/Makefile]: CMAKE_INSTALL takes care of this. No need to this section.

File: libs/h2o/Makefile [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  URL:=https://h2o.examp1e.net/
+  DEPENDS:=+libuv +libwslay +libopenssl +zlib +libyaml +ruby
+endef
+
+define Package/libh2o
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=H2O Library compiled with libuv
+  URL:=https://h2o.examp1e.net/
+  DEPENDS:=+libuv +libwslay +libopenssl +zlib +libh2o-evloop +libyaml
```
Comment (neheb) [inline libs/h2o/Makefile]: This can be replaced with +libh2o-evloop

File: libs/h2o/Makefile [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+PKG_LICENSE_FILES:=LICENSE
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+PKG_BUILD_DEPENDS:=ruby/host
+PKG_BUILD_PARALLEL:=1
+
+CMAKE_OPTIONS:= \
+	-DCMAKE_INSTALL_PREFIX=/usr \
```
Comment (neheb) [inline libs/h2o/Makefile]: These two are already default.

File: libs/h2o/Makefile (hunk 1 above)
Comment (James-TR) [reply libs/h2o/Makefile]: Removed :)

File: libs/h2o/Makefile (hunk 3 above)
Comment (James-TR) [reply libs/h2o/Makefile]: Removed :)

File: libs/h2o/Makefile (hunk 2 above)
Comment (James-TR) [reply libs/h2o/Makefile]: I've removed the dependency on libuv from libh2o-evloop as well, so now libh2o only depends on libuv and evloop. I'm not sure if it actually depends on libh2o-evloop or not but that's how the debian packages do their dependencies.

**Comment by James-TR** [comment]
> I suggest splitting your single commit into multiple ones. It's preferred to have one single commit per one new package.
> 
> //EDIT: In your commit, there are removed three patches, but not mentioned anywhere.

I've documented the removal of compatibility patches in the commit for dnsdist, and also split everything out into its own commit

File: libs/h2o/Makefile (hunk 2 above)
Comment (James-TR) [reply libs/h2o/Makefile]: I've corrected the dependencies after reviewing the debian packages. The two implementations are completely separate, the only differentiating factor is that libh2o uses libuv for its event handler, and libh2o-evloop implements its own.

**Comment by neheb** [comment]
h2o it seems has some issues:

```
[100%] Building C object CMakeFiles/h2o.dir/deps/picotls/lib/openssl.c.o
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c: In function 'cipher_encrypt':
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:422:26: warning: unused variable 'ret' [-Wunused-variable]
  422 |     int len = (int)_len, ret = EVP_EncryptUpdate(ctx->evp, output, &len, input, len);
      |                          ^~~
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c: In function 'chacha20_setup_crypto':
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:440:42: warning: implicit declaration of function 'EVP_chacha20'; did you mean 'LN_chacha20'? [-Wimplicit-function-declaration]
  440 |     return cipher_setup_crypto(ctx, key, EVP_chacha20(), cipher_encrypt);
      |                                          ^~~~~~~~~~~~
      |                                          LN_chacha20
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:440:42: warning: passing argument 3 of 'cipher_setup_crypto' makes pointer from integer without a cast [-Wint-conversion]
  440 |     return cipher_setup_crypto(ctx, key, EVP_chacha20(), cipher_encrypt);
      |                                          ^~~~~~~~~~~~~~
      |                                          |
      |                                          int
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:400:96: note: expected 'const EVP_CIPHER *' {aka 'const struct evp_cipher_st *'} but argument is of type 'int'
  400 | static int cipher_setup_crypto(ptls_cipher_context_t *_ctx, const void *key, const EVP_CIPHER *cipher,
      |                                                                              ~~~~~~~~~~~~~~~~~~^~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c: In function 'aead_chacha20poly1305_setup_crypto':
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:590:48: warning: implicit declaration of function 'EVP_chacha20_poly1305'; did you mean 'LN_chacha20_poly1305'? [-Wimplicit-function-declaration]
  590 |     return aead_setup_crypto(ctx, is_enc, key, EVP_chacha20_poly1305());
      |                                                ^~~~~~~~~~~~~~~~~~~~~
      |                                                LN_chacha20_poly1305
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:590:48: warning: passing argument 4 of 'aead_setup_crypto' makes pointer from integer without a cast [-Wint-conversion]
  590 |     return aead_setup_crypto(ctx, is_enc, key, EVP_chacha20_poly1305());
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~
      |                                                |
      |                                                int
/home/mangix/devstuff/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/h2o-2.2.6/deps/picotls/lib/openssl.c:531:104: note: expected 'const EVP_CIPHER *' {aka 'const struct evp_cipher_st *'} but argument is of type 'int'
  531 | static int aead_setup_crypto(ptls_aead_context_t *_ctx, int is_enc, const void *key, const EVP_CIPHER *cipher)
      |                                                                                      ~~~~~~~~~~~~~~~~~~^~~~~~
[100%] Linking CXX executable h2o
/home/mangix/devstuff/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-9.2.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/9.2.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: CMakeFiles/h2o.dir/deps/picotls/lib/openssl.c.o: in function `aead_chacha20poly1305_setup_crypto':
h2o-2.2.6/h2o-2.2.6/deps/picotls/lib/openssl.c:590: undefined reference to `EVP_chacha20_poly1305'
/home/mangix/devstuff/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-9.2.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/9.2.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: CMakeFiles/h2o.dir/deps/picotls/lib/openssl.c.o: in function `chacha20_setup_crypto':
h2o-2.2.6/h2o-2.2.6/deps/picotls/lib/openssl.c:440: undefined reference to `EVP_chacha20'
```

I imagine it's a simple #ifdef to check if chacha20 is available.

**Comment by James-TR** [comment]
if !defined(OPENSSL_NO_CHACHA) && !defined(OPENSSL_NO_POLY1305) might be suitable to guard against that. There's a header which exposes it :)

**Comment by James-TR** [comment]
I've opened https://github.com/h2o/h2o/pull/2195 and https://github.com/h2o/picotls/pull/285 with the patches against h2o

**Comment by James-TR** [comment]
Closed the PR by accident... woops

**Comment by James-TR** [comment]
Is there a way to force this to build with -j1 just for circleci? the lua compilation in dnsdist and pdns-recursor can use up a tonne of ram

**Comment by neheb** [comment]
Same with boost. I'd just ignore it for now. The buildbots will tell if it fails.

**Comment by James-TR** [comment]
Thanks!

**Comment by neheb** [comment]
Anyway, looks good. Merging.

**Pull request merged**